src/mlpack/tests/augmented_rnns_tasks_test.cpp

```diff
@@ -0,0 +1,291 @@
+#include <mlpack/methods/ann/layer/leaky_relu.hpp>
+#include <mlpack/methods/ann/rnn.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using std::vector;
+using std::pair;
+using std::make_pair;
+
+using namespace mlpack::ann::augmented;
+using namespace mlpack::ann::augmented::tasks;
+using namespace mlpack::ann::augmented::scorers;
+
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+
+using mlpack::data::Binarize;
+
+// The dummy model that simply copies the sequence
+// the required number of times
+// (yes, no ML here, we're unit testing :)
+class HardCodedCopyModel
+{
+ public:
+  HardCodedCopyModel() : nRepeats(1) {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    arma::mat input = predictors.at(0);
+    arma::mat output = labels.at(0);
+    size_t zeroCnt = 0, oneCnt = 0;
+    for (size_t i = 1; i < input.n_rows; i += 2)
+    {
+      size_t& addVar = (input.at(i, 0) == 0) ? zeroCnt : oneCnt;
+      ++addVar;
+    }
+    assert(oneCnt % zeroCnt == 0);
+    nRepeats = oneCnt / zeroCnt;
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    size_t seqLen = (predictors.n_rows / 2) / (nRepeats + 1);
+    size_t outputLen = nRepeats * seqLen;
+    assert(2 * (seqLen + outputLen) == predictors.n_rows);
+    labels.zeros(predictors.n_rows / 2, 1);
+    for (size_t i = 0; i < outputLen; ++i)
+    {
+      labels.at(seqLen+i) = predictors.at(2 * (i % seqLen));
+    }
+  }
+  void Predict(arma::field<arma::mat>& predictors,
+               arma::field<arma::mat>& labels)
+  {
+    size_t sz = predictors.n_elem;
+    labels = arma::field<arma::mat>(sz);
+    for (size_t i = 0; i < sz; ++i)
+    {
+      Predict(predictors.at(i), labels.at(i));
+    }
+  }
+
+ private:
+  size_t nRepeats;
+};
+
+// The dummy model that simply sorts the sequence.
+class HardCodedSortModel {
+ public:
+  HardCodedSortModel(size_t bitLen) : bitLen(bitLen) {}
+  void Train(arma::field<arma::mat>& predictors,
+             arma::field<arma::mat>& labels)
+  {
+    assert(predictors.n_elem == labels.n_elem);
+  }
+  void Predict(arma::mat& predictors,
+               arma::mat& labels)
+  {
+    predictors = predictors.t();
+    predictors.reshape(bitLen, predictors.n_elem / bitLen);
+    size_t len = predictors.n_cols;
+    labels.zeros(bitLen, len);
+    vector<pair<int, int>> vals(len);
+    for (size_t j = 0; j < len; ++j)
+    {
+      int val = 0;
+      for (size_t k = 0; k < bitLen; ++k)
+      {
+        val <<= 1;
+        val += predictors.at(k, j);
+      }
+      vals[j] = make_pair(val, j);
+    }
+    sort(vals.begin(), vals.end());
+    for (size_t j = 0; j < len; ++j)
+    {
+      labels.col(j) = predictors.col(vals[j].second);
+    }
+    labels.reshape(predictors.n_elem, 1);
+  }
+  void Predict(arma::field<arma::mat>& predictors,
+               arma::field<arma::mat>& labels)
+  {
+    auto sz = predictors.n_elem;
```

Comment:
I think we can directly use ``predictors.n_elem;`` instead of ``sz`` here. In general we should avoid to use auto when using armadillo:

> Use of C++11 auto is not recommended with Armadillo objects and expressions. Armadillo has a template meta-programming framework which creates lots of short lived temporaries that are not handled by auto.

In this case it would be just fine, but I think using it directly without using an alias is also just fine, an we can save another line.